sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeWriter.java

```diff
@@ -17,17 +17,86 @@
 package org.apache.spark.sql.catalyst.expressions.codegen;
 
 import org.apache.spark.sql.types.Decimal;
+import org.apache.spark.unsafe.Platform;
+import org.apache.spark.unsafe.array.ByteArrayMethods;
 import org.apache.spark.unsafe.types.CalendarInterval;
 import org.apache.spark.unsafe.types.UTF8String;
 
 /**
  * Base class for writing Unsafe* structures.
  */
 public abstract class UnsafeWriter {
+  // Keep internal buffer holder
+  protected final BufferHolder holder;
+
+  // The offset of the global buffer where we start to write this structure.
+  protected int startingOffset;
+
+  protected UnsafeWriter(BufferHolder holder) {
+    this.holder = holder;
+  }
+
+  /**
+   * Accessor methods are delegated from BufferHolder class
+   */
+  public final BufferHolder getBufferHolder() {
+    return holder;
+  }
+
+  public final byte[] buffer() {
+    return holder.buffer();
+  }
+
+  public final void reset() {
+    holder.reset();
+  }
+
+  public final int totalSize() {
+    return holder.totalSize();
+  }
+
+  public final void grow(int neededSize) {
+    holder.grow(neededSize);
+  }
+
+  public final int cursor() {
+    return holder.getCursor();
+  }
+
+  public final void incrementCursor(int val) {
+    holder.incrementCursor(val);
+  }
+
+  public abstract void setOffsetAndSizeFromMark(int ordinal, int mark);
```

Comment:
`Mark` is an ambiguous term. It is not clear what it means here.

Reply:
Btw, why we have `_setOffsetAndSizeFromMark` and `setOffsetAndSizeFromMark`? Seems `setOffsetAndSizeFromMark` just call `_setOffsetAndSizeFromMark`.

Reply:
Why do we even have this method? I feel it's cleaner if the caller side just do `writer.setOffsetAndSize(ordinal, tmpCursor, currentCursor - tmpCursor);`

Reply:
I don't actually. I think this should be the responsibility of the writer, and we are doing the same thing in three different places.

Reply:
But "mark" is really confusing, how about `setOffsetAndSizeFromPreviousCursor`?